File: 56. Merge Intervals/solution.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.size() == 1 || intervals.size() == 0) { return intervals; }
        
        std::sort(intervals.begin(), intervals.end(), [](vector<int>& left, vector<int>& right) {
            return left[0] < right[0];
        });
        
        vector<vector<int>> result;
        result.push_back(intervals[0]);
        int i = 1;
        while (i < intervals.size()) {
            while (i < intervals.size() && intervals[i][0] <= result[result.size() - 1][1]) {
                result[result.size() - 1][1] = max(intervals[i][1], result[result.size() - 1][1]);
                i++;
            }
            if (i < intervals.size()) {
                result.push_back(intervals[i++]);
            }
        }
        return result;
    }
};

File: 56. Merge Intervals/solution.java
class Solution {
    class Compare implements Comparator<int[]> {
        public int compare(int[] a, int[] b) {
            return a[0] - b[0];
        }
    }
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, new Compare());
        int i = 0;
        int probe = i;
        List<Integer[]> listResult = new ArrayList<>();
        while (probe < intervals.length) {
            int start = intervals[i][0];
            int end = intervals[i][1];
            while (probe < intervals.length && intervals[probe][0] <= end) {
                start = Math.min(start, intervals[probe][0]);
                end = Math.max(end, intervals[probe][1]);
                probe++;
            }
            
            Integer[] newInterval = new Integer[2];
            newInterval[0] = start;
            newInterval[1] = end;
            // System.out.println("start: " + start + " end: " + end);
            listResult.add(newInterval);
            // result[resultIndex++] = newInterval;
            i = probe;
        }
        int resultIndex = 0;
        int[][] result = new int[listResult.size()][2];
        while (resultIndex < listResult.size()) {
            result[resultIndex][0] = listResult.get(resultIndex)[0];
            result[resultIndex][1] = listResult.get(resultIndex)[1];
            resultIndex++;
        }
        return result;
    }
}
